### ISSUE TITLE
"Failed" replies may actually succeed after client exits

### ISSUE THREAD

**Issue by eloquence** [freedomofpress/securedrop-client#1457]
# Steps to reproduce
1. On your test server, create a new source and stay logged in as that source. Send a message.
2. Log into the app and wait for it to sync to pick up the new source.
3. Disconnect sd-whonix from `sys-firewall`.
4. Send a reply to the new source
5. Log out (reply transitions to failed)
6. Exit the app completely
7. Reconnect sd-whonix to `sys-firewall` (within 60 seconds or so of exiting the app)
8. Reload the source session to check for replies

# Expected behavior

No reply

# Actual behavior

1. After a while, the reply happily arrives in the Source Interface. 
2. Next time you log into the Client, it is still initially marked as "failed", but after the next sync, the client detects and downloads the reply.

**Comment by eloquence** [comment]
Observed with `securedrop-client_0.6.0-dev-20220331-060621` on a workstation provisioned with the Qubes 4.1 PR (https://github.com/freedomofpress/securedrop-workstation/pull/751), so a bit of an odd setup, but I'm pretty sure I've observed this edge case before without realizing it. If you are unable to reproduce it, I can re-test on a more prod-like environment.

I'm guessing what's happening here is that the network call has been dispatched and will eventually succeed, unrelated to any queuing shenanigans. I'm not sure what _should_ happen here -- do we want to attempt to kill requests that are in progress? It might argue for not setting the "failed" state until we truly know that a reply hasn't made it to the server.

**Comment by cfm** [comment]
Thanks for the report, @eloquence: detailed, relevant, and timely.

> I am also filing it as an edge to consider in verifying #1420, since it seems like a distinct issue.

Up to a point!  Currently they are distinct cases of undefined behavior at different layers of the application: here, at the network level outbound from `securedrop-proxy`; in (e.g.) <https://github.com/freedomofpress/securedrop-client/pull/1434#issuecomment-1084098646>, at the job level under processing in `securedrop-client`.  I confess I'm feeling another state machine coming on: the lifecycle of an `ApiJob` from `securedrop-client` → `securedrop-proxy` → `sd-whonix` → SecureDrop and back again....

**Comment by rocodes** [comment]
Just commenting since I have also seen this that the GUI behaviour here includes a reply moving from the "failed" state to the "successful" state (i.e. colour change), but retaining the pink "Failed" label beside the reply (confirmed with @eloquence that he sees the same thing).

**Comment by cfm** [comment]
[Update during SecureDrop Workstation hangout:]  I misspoke above in <https://github.com/freedomofpress/securedrop-client/issues/1457#issuecomment-1085452092>.  My goal here is not a state diagram but one or more sequence diagrams, illustrating that (a) the Server is authoritative for the set of successful replies but (b) the Client believes itself to be authoritative for the state of replies it's originated.  That conflict in this case may prove to require only a UI fix, but it has conceptual ramifications about how we (should) reason about synchronization between the Client and the Server that are worth documenting.

**Comment by creviera** [comment]
I would have to look at this again, but it seems like the way we auto-fail pending replies when you switch into offline mode might not have been the correct design choice. We should consider just leaving replies in a pending state while in offline mode and when you first log in.

**Comment by rocodes** [comment]
>  We should consider just leaving replies in a pending state while in offline mode and when you first log in.

Not to stir the pot, but I feel a little hesitant about this.... I feel like there are certain retry patterns that could help here if we don't want to immediately fail replies during offline mode (eg https://en.wikipedia.org/wiki/Exponential_backoff with a max number of retries), but it makes sense to me that if the user signs out and chooses offline mode (vs a temporary network disruption), replies should fail.

**Comment by creviera** [comment]
It's definitely not stirring the pot, but first, I want to add context to what I'm saying. There are two cases I'm thinking about:

#### Case 1
1. Journalist sends a reply
2. Reply doesn't make it to the server and no server response is received because the journalist switches to offline mode or exits

#### How the client currently handles this

The reply shows up as failed when the user logs out. The reply continues to show up as failed when the user logs in.

#### What do we expect?

For this case, the behavior looks correct since the reply never made it to the server. But assuming the reply never made it is problematic (see case 2 below). But this is just out of luck, because the client doesn't know if it actually failed.

#### Case 2
1. Journalist sends a reply
2. Reply **does** make it to the server and no server response is received because the journalist switches to offline mode or exits
3. User goes back online

#### How the client currently handles this

The reply shows up as failed when the user logs out. The reply switches from failed to succeeded when the user logs in.

#### What do we expect?

Here's what I would expect:

The reply stays pending when the user logs out. The reply switches to succeeded when the user logs in.

> Not to stir the pot, but I feel a little hesitant about this.... I feel like there are certain retry patterns that could help here if we don't want to immediately fail replies during offline mode (eg https://en.wikipedia.org/wiki/Exponential_backoff with a max number of retries), but it makes sense to me that if the user signs out and chooses offline mode (vs a temporary network disruption), replies should fail.

Since we don't want the client's retry logic to execute after the user logs out, why not continue to show the reply as pending until the client knows whether or not it made it to the server once the user logs in again?

**Comment by rocodes** [comment]
Right, that distinction makes a lot of sense, thank you @creviera. I agree with you with the desired behaviour in these two cases. (And I will wait for Cory's reply since he mentioned he also wants to elaborate).

**Comment by cfm** [comment]
@creviera in <https://github.com/freedomofpress/securedrop-client/issues/1457#issuecomment-1121782720>:

> I would have to look at this again, but it seems like the way we auto-fail pending replies when you switch into offline mode might not have been the correct design choice. We should consider just leaving replies in a pending state while in offline mode and when you first log in.

Thanks again for this suggestion and our conversation today, which have unlocked this bug for me.  It turns out to have nothing to do with how API jobs are enqueued, processed, and retried.  Instead:

1. On login, on logout, and on switching to offline mode, the GUI calls the self-explanatory `securedrop_client.storage.mark_all_pending_draft_replies_as_failed()`.
2. However, at any of these state transitions, the `SendReplyJob`'s underlying `POST` request may have already been made and is just waiting to go through on the network—even if the application has quit by then!

Your suggestion here lets us resolve this bug by decoupling the GUI-level state transition from the status of any particular `DraftReply`.

However, this resolution emphasizes the point @creviera and I discussed today:  Right now the concept of "draft replies", as opposed to a single object with multiple statuses, is the lone special case where we distinguish between *completion of the user's intent* (the pending or "draft" reply) and the *successful execution of a request or command* (the `SendReplyJob`).  That's the distinction I've been arguing we need to design for generally: first in order to able to reason about failure modes like this one for *all* user-initiated actions; maybe in the future in order to be able to support a first-class offline mode.

Here's a (rough, whiteboardish) [slide](https://docs.google.com/presentation/d/1aSyH0oDHh6g0n9Wm8BQsRaFUSf6vWf8kP2PdHcc1nF0/edit#slide=id.p) @creviera and I used today to discuss how to extrapolate from this `SendReplyJob` special case.  I can also sequence-diagram it if that would be helpful.

For now, draft pull request forthcoming, with architectural next steps to follow separately.
